prototypes/ScheduledMerges.hs

```diff
@@ -646,63 +909,271 @@ tieringLevelIsFull _ln _incoming resident = length resident >= 4
+            state' <- OngoingTreeMerge <$> newLeafMerge nullTracer mergeType rs
+            -- use any remaining credits to progress the new merge
+            supplyCreditsMergingTreeState c' state'
+
+supplyCreditsPendingMerge :: Credit -> PendingMerge s -> ST s Credit
+supplyCreditsPendingMerge = checked remainingDebtPendingMerge $ \credits pm -> do
+    let PendingMerge mergeType incoming trees = pm
+    assertST (credits >= 0)
+    case mergeType of
+      MergeLevel _ -> leftToRight credits incoming trees
+      MergeUnion ->
+        case (incoming, trees) of
+          ([], [t1, t2]) -> splitEqually credits t1 t2
+          _              -> error $ "supplyCreditsPendingMerge: "
+                                 ++ "expected two union merge inputs, "
+                                 ++ "got " ++ show (length trees)
+  where
+    -- supply credit left to right until it is used up
+    leftToRight 0 _         _       = return 0
+    leftToRight c []       []       = return c
+    leftToRight c []       (mt:mts) = do
+        c' <- supplyCreditsMergingTree c mt
+        leftToRight c' [] mts
+    leftToRight c (ir:irs) mts      = do
+        c' <- case ir of
+          Single _     -> return c
+          Merging _ mr -> supplyCreditsMergingRun c mr
+        leftToRight c' irs mts
+
+    splitEqually c mt1 mt2 = do
+        -- supply credit roughly evenly on both sides
+        let (c1, c2) = (c `div` 2, c - c1)
+        c1' <- supplyCreditsMergingTree c1 mt1
+        if c1' > 0
+          then
+            -- left side done, use all remaining credits on the right side
+            supplyCreditsMergingTree (c2 + c1') mt2
+          else do
+            -- left side still not done
+            -- if the right side has leftovers, go back
+            c2' <- supplyCreditsMergingTree c2 mt2
+            if c2' > 0
+              then supplyCreditsMergingTree c2' mt1
+              else return 0
+
+-- TODO: Tracer
+expectCompletedChildren :: HasCallStack => PendingMerge s -> ST s (MergeType, [Run])
+expectCompletedChildren (PendingMerge mergeType irs trees) = do
+    rs1 <- traverse expectCompletedIncomingRun irs
+    rs2 <- traverse expectCompletedMergingTree trees
+    return (mergeType, rs1 ++ rs2)
+  where
+    expectCompletedIncomingRun = \case
+        Single     r -> return r
+        Merging _ mr -> expectCompletedMerge nullTracer Nothing mr
+
+expectCompletedMergingTree :: HasCallStack => MergingTree s -> ST s Run
+expectCompletedMergingTree (MergingTree ref) = do
+    readSTRef ref >>= \case
+      CompletedTreeMerge r    -> return r
+      OngoingTreeMerge _ -> error $ "expectCompletedMergingTree: OngoingTreeMerge"
+      PendingTreeMerge _ -> error $ "expectCompletedMergingTree: PendingTreeMerge"
 
 -------------------------------------------------------------------------------
 -- Measurements
 --
 
-allLayers :: LSM s -> ST s [[Run]]
+-- TODO: Is this useful or should we directly flatten to a Map?
```

Comment:
It's a question I'd like to answer before merging. It's not hard to do either way, but I'm currently unsure on what makes more sense.

Reply:
So `MTree` is only returned by `allLayers` which is used only in `logicalValue` where the `MTree` is merged down to a single `Run`.

I'm not sure what you mean about flattening to a `Map`. The flattening that happens in `logicalValue` is `mergeTree` which works recursively.

Reply:
I think with `Map` I meant a run (as that's represented as a map). So instead of creating an `MTree` which later gets merged down to a run, just do the merging right away. 

However, the `MTree` might still be useful e.g. for making assertiongs about the tree shape in tests (which goes via `dumpRepresentation`, see https://github.com/IntersectMBO/lsm-tree/pull/477#discussion_r1862306609).